elements/lisk-p2p/src/directory/new_peers.ts

```diff
@@ -0,0 +1,271 @@
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+		if (bucket && bucket.get(incomingPeerId)) {
+			const success = bucket.delete(incomingPeerId);
+			this._newPeerMap.set(bucketId, bucket);
+
+			return success;
+		}
+
+		return false;
+	}
+
+	public getPeer(peerInfo: P2PPeerInfo): P2PPeerInfo | undefined {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+		if (bucket) {
+			const newPeer = bucket.get(incomingPeerId);
+
+			return newPeer ? newPeer.peerInfo : undefined;
+		}
+
+		return undefined;
+	}
+
+	public addPeer(peerInfo: P2PPeerInfo): AddPeerOutcome {
+		const bucketId = this.getBucketId(peerInfo.ipAddress);
+		const bucket = this._newPeerMap.get(bucketId);
+		const incomingPeerId = constructPeerIdFromPeerInfo(peerInfo);
+
+		if (bucket && bucket.get(incomingPeerId)) {
+			return {
+				success: false,
+				isEvicted: false,
+			};
+		}
+
+		if (bucket) {
+			const newTriedPeerInfo = {
+				peerInfo,
+				numOfConnectionFailures: 0,
+				dateAdded: new Date(),
+			};
+
+			if (bucket.size < this._newPeerBucketSize) {
+				bucket.set(incomingPeerId, newTriedPeerInfo);
+				this._newPeerMap.set(bucketId, bucket);
+
+				return {
+					success: true,
+					isEvicted: false,
+				};
+			} else {
+				const evictedPeer = this._evictPeer(bucketId);
+				bucket.set(incomingPeerId, newTriedPeerInfo);
+				this._newPeerMap.set(bucketId, bucket);
+
+				return {
+					success: true,
+					isEvicted: true,
+					evictedPeer: evictedPeer.peerInfo,
+				};
+			}
+		}
+
+		return {
+			success: false,
+			isEvicted: false,
+		};
+	}
+
+	public failedConnectionAction(incomingPeerInfo: P2PPeerInfo): boolean {
+		const success = this.removePeer(incomingPeerInfo);
+
+		return success;
+	}
+
+	private _evictPeer(bucketId: number): NewPeerInfo {
+		const peerList = this._newPeerMap.get(bucketId);
+
+		if (!peerList) {
+			throw new Error(`No Peer list for bucket Id: ${bucketId}`);
+		}
+
+		// First eviction strategy
+		const evictedPeerBasedOnTime = this._evictionBasedOnTimeInBucket(
+			bucketId,
+			peerList,
+		);
+
+		if (evictedPeerBasedOnTime) {
+			return evictedPeerBasedOnTime;
+		}
+
+		// Second eviction strategy
+		return this._evictionRandom(bucketId);
+	}
+
+	private _evictionBasedOnTimeInBucket(
+		bucketId: number,
+		peerList: Map<string, NewPeerInfo>,
+	): NewPeerInfo | undefined {
+		// tslint:disable-next-line:no-let
+		let evictedPeer: NewPeerInfo | undefined;
+
+		[...this._newPeerMap.values()].forEach(peersMap => {
+			[...peersMap.keys()].forEach(peerId => {
+				const peer = peersMap.get(peerId);
+				if (peer) {
+					const diffDays = Math.round(
+						Math.abs(
+							(peer.dateAdded.getTime() - new Date().getTime()) /
+								MILLISECONDS_IN_ONE_DAY,
+						),
+					);
+					if (diffDays >= ELIGIBLE_DAYS_FOREVICTION) {
```

Comment:
Ideally this should be customizable from the top level P2P class; the current constant value can be the default. That way later we can simulate the effects of this feature in fast forward in integration tests.

This would affect the `MILLISECONDS_IN_ONE_DAY` constant above too.

Reply:
I think it will not affect `MILLISECONDS_IN_ONE_DAY` because that is used to just return the difference between two dates in days.